### PR TITLE
[ENG-6057][cli] api key creation: exponential backoff for slow apple propagation

### PR DESCRIPTION
**Pull request by quinlanj**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why
There is a [bug in Apple's infrastructure](https://developer.apple.com/forums/thread/684314) that does not propagate newly created objects for a long time (up to 14 seconds in my experiments!). This behaviour is non-deterministic, but the general chain of events is as follows:
1. cli creates Api Key through Apple API
2. cli receives response that key is created
3. cli attempts to download Api Key
4. Apple responds saying that resource does not exist 😿 

Fixes https://github.com/expo/expo-cli/issues/4521

# How
Screen Shot 2022-08-22 at 9 38 40 PM

Retry, using exponential backoff. We will retry for up to 6 times, waiting 2, 4, 8, 16, 32 and 64 seconds between each retry. I tried creating an Api Key many times and when the bug occurred, the maximum time I ended up waiting was around 14 seconds. It is possible people might experience longer times, so I added more retries just in case (hopefully waiting 64 seconds in the worst case scenario is not too awful!). 

In the case where the key never propagates, I fail with a helpful error message showing people how to create their key manually. I've also added analytics to track the types of failures and the # of retries users experience - this propagation issue is new, and Apple's server side code may change to eventually fix this.

# Test Plan

- [x] new tests pass

### CODE REVIEW

File: packages/eas-cli/src/analytics/events.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ export enum SubmissionEvent {
   SUBMIT_REQUEST_ATTEMPT = 'submit cli request attempt',
   SUBMIT_REQUEST_SUCCESS = 'submit cli request success',
   SUBMIT_REQUEST_FAIL = 'submit cli request fail',
+  CREDENTIALS_API_KEY_DOWNLOAD = 'submit cli credentials api key download',
```
Comment (wkozyra95) [inline packages/eas-cli/src/analytics/events.ts:18]: It looks like you are registering this even only on failure, so
```suggestion
  CREDENTIALS_API_KEY_DOWNLOAD_FAIL = 'submit cli credentials api key download fail',
```
and maybe a success event with a final number of retries would also be useful.

Also, maybe consider creating a separate group e.g. `CredentialsEvent`

**Review by wkozyra95 — APPROVED** [review]

File: packages/eas-cli/src/credentials/ios/appstore/ascApiKey.ts [hunk 2]
```diff
@@ -49,6 +51,70 @@ export async function getAscApiKeyAsync(
...
+ * the resource does not exist. We retry with exponential backoff until the key propagates and
+ * is available for download.
+ * */
+export async function downloadWithRetryAsync(
+  key: ApiKey,
+  options: { minTimeout?: number } = {}
+): Promise<string | null> {
+  const retries = 6;
+  const factor = 2;
+  const minTimeout = options.minTimeout ?? 1000;
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/appstore/ascApiKey.ts:66]:
```suggestion
 { minTimeout = 1000, retries = 6, factor = 2 } = {}
): Promise<string | null> {
```

File: packages/eas-cli/src/credentials/ios/appstore/ascApiKey.ts [hunk 3]
```diff
@@ -49,6 +51,70 @@ export async function getAscApiKeyAsync(
...
+        minTimeout,
+      }
+    );
+  } catch (e: any) {
+    if (
+      e.name === 'UnexpectedAppleResponse' &&
+      e.message.includes(RESOURCE_DOES_NOT_EXIST_MESSAGE)
+    ) {
+      Log.warn(
+        `Unable to download Api Key from Apple at this time. Please create and upload your key manually by running 'eas credentials' ${learnMore(
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/appstore/ascApiKey.ts:105]: Please remove "please". Please see https://github.com/expo/eas-cli/pull/1250, please.

**Review by dsokal — APPROVED** [review]
Add a changelog entry.

**Comment by quinlanj** [comment]
/changelog-entry bug-fix Retry ASC Api Key Downloads if it has not fully propagated on Apple's infrastructure.

**Pull request merged**